motionplan/check.go

```diff
@@ -0,0 +1,430 @@
+package motionplan
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/motionplan/ik"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+// CheckPlan checks if obstacles intersect the trajectory of the frame following the plan. If one is
+// detected, the interpolated position of the rover when a collision is detected is returned along
+// with an error with additional collision details.
+func CheckPlan(
+	checkFrame referenceframe.Frame, // TODO(RSDK-7421): remove this
+	executionState ExecutionState,
+	worldState *referenceframe.WorldState,
+	fs referenceframe.FrameSystem,
+	lookAheadDistanceMM float64,
+	logger logging.Logger,
+) error {
+	plan := executionState.Plan()
+	startingInputs := plan.Trajectory()[0]
+	wayPointIdx := executionState.Index()
+
+	// ensure that we can actually perform the check
+	if len(plan.Path()) < 1 {
+		return errors.New("plan must have at least one element")
+	}
+	if len(plan.Path()) <= wayPointIdx || wayPointIdx < 0 {
+		return errors.New("wayPointIdx outside of plan bounds")
+	}
+
+	// construct solverFrame
+	// Note that this requires all frames which move as part of the plan, to have an
+	// entry in the very first plan waypoint
+	sf, err := newSolverFrame(fs, checkFrame.Name(), referenceframe.World, startingInputs)
+	if err != nil {
+		return err
+	}
+	// construct planager
+	sfPlanner, err := newPlanManager(sf, logger, defaultRandomSeed)
+	if err != nil {
+		return err
+	}
+	// This should be done for any plan whose configurations are specified in relative terms rather than absolute ones.
+	// Currently this is only TP-space, so we check if the PTG length is >0.
+	// The solver frame will have had its PTGs filled in the newPlanManager() call, if applicable.
+	if sfPlanner.useTPspace {
+		return checkPlanRelative(checkFrame, executionState, worldState, fs, lookAheadDistanceMM, sfPlanner)
+	}
+	return checkPlanAbsolute(checkFrame, executionState, worldState, fs, lookAheadDistanceMM, sfPlanner)
+}
+
+func checkPlanRelative(
+	checkFrame referenceframe.Frame, // TODO(RSDK-7421): remove this
+	executionState ExecutionState,
+	worldState *referenceframe.WorldState,
+	fs referenceframe.FrameSystem,
+	lookAheadDistanceMM float64,
+	sfPlanner *planManager,
+) error {
+	plan := executionState.Plan()
+	startingInputs := plan.Trajectory()[0]
+	currentInputs := executionState.CurrentInputs()
+	currentPoseIF := executionState.CurrentPoses()[checkFrame.Name()]
+	wayPointIdx := executionState.Index()
+	sf := sfPlanner.frame
+
+	// Validate the given PoseInFrame of the relative frame. Relative frame poses cannot be given in their own frame, or the frame of
+	// any of their children.
+	// TODO(RSDK-7421): there will need to be checks once there is a real possibility of multiple, hierarchical relative frames, or
+	// something expressly forbidding it.
+	validateRelPiF := func(pif *referenceframe.PoseInFrame) error {
```

Comment:
Since this function is now in its own file I would not be opposed to moving `validateRelPiF` and `toWorld` outside of this function and into their own non-anonymous function. 
I understand that these functions are only called by `checkPlanRelative` so it does make sense to keep them here as is, but I feel that it might hurt readability.

Reply:
They're currently helper functions that make heavy use of the scope in which they exist; moving them outside that scope would drastically increase the number of parameters required and probably hurt readability just as much.